### PR TITLE
Implement delete fhir endpt logic

### PR DESCRIPTION
**Pull request by karenzheung**
This PR implements the logic for deleting fhir endpoints based on update times following M2/K1 as described here:
https://mitre.sharepoint.com/sites/ProjectLantern/_layouts/15/Doc.aspx?sourcedoc={bc496df4-9591-43a5-8860-8828979cb683}&action=edit&wd=target%28Development%20Plan.one%7C273069de-a04d-41e2-9f83-5cb18ce5d5e5%2FMaintaining%20list%20of%20current%20endpoints%7C56dac67e-e104-3741-ba05-14571baeed01%2F%29

To test:
This is easier to test if use just the CareEvolution list source and comment out the others. There will still be some endpoints from NPPES.  
1. run `make populatedb`
2. `cd endpointmanager/cmd` and run `go run sendendpoints/main.go`
3. In same directory but different terminal run `go run capabilityreceiver/main.go`
4. Check the database `docker exec -it lantern-back-end_pg_prometheus_1 psql -U lantern`
and verify that fhir_endpoints and fhir_endpoints_info have entries in them. Note the updated_at time of the endpoints and the number of endpoints in each table. Also check endpoints that are in the table multiple times from different list source.
`select * from fhir_endpoints where url in (select url from fhir_endpoints group by url having count(*) > 1) order by url;`
5. edit the populatedb.sh and replace the line
`go run main.go ../../../networkstatsquerier/resources/EndpointSources.json CareEvolution`
with:
`go run main.go ../../../networkstatsquerier/resources/EndpointSources_1.json CareEvolution`
6. run `make populatedb.sh`
7. Check the database and verify that there are 10 less endpoints in fhir_endpoints. fhir_endpoints_info table will have 6 less endpoints because of the 4 of the 10 endpoints that were removed in EndpointSources_1.json still exist from the NPPES list source. Check the fhir_endpoints_info_history table to see the deleted endpoints. 
`select operation, url, updated_at from fhir_endpoints_info_history where operation='D';`

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-198
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

### CODE REVIEW

File: endpointmanager/pkg/fhirendpointquerier/populatefhirendpoints_integration_test.go [hunk 1]
```diff
@@ -216,6 +237,53 @@ func Test_AddEndpointData(t *testing.T) {
 		fmt.Sprintf("stored data '%v' does not equal expected store data '%v'", storedEndpt.OrganizationNames, endpt2.OrganizationNames))
 }
 
+func Test_RemoveOldEndpoints(t *testing.T) {
```
Comment (bmath10) [inline endpointmanager/pkg/fhirendpointquerier/populatefhirendpoints_integration_test.go:240]: I think this is a good test but it doesn't actually call `removeOldEndpoints` so it kind of seems like additional test cases to `Test_AddEndpointData` than a test for `removeOldEndpoints`. I think this test should get merged into the `Test_AddEndpointData` and the `Test_RemoveOldEndpoints` should be something like
1. Add an endpoint
2. Add a second endpoint
3. Get the updated_at time for the second endpoint
4. Call removeOldEndpoints using that time and make sure the first endpoint got deleted.

I think it probably makes sense to also check the case that if you have an endpoint with that URL but a different list source and there is an entry in the `fhir_endpoint_info` table with that URL that you don't delete the fhir_endpoint_info entry

File: endpointmanager/pkg/fhirendpointquerier/populatefhirendpoints_integration_test.go [hunk 2]
```diff
@@ -216,6 +237,53 @@ func Test_AddEndpointData(t *testing.T) {
...
+	th.Assert(t, err == nil, err)
+	err = ctStmt.QueryRow().Scan(&ct)
+	th.Assert(t, err == nil, err)
+	th.Assert(t, ct == expectedEndptsStored, fmt.Sprintf("Expected %d products stored. Actually had %d products stored.", expectedEndptsStored, ct))
+	storedEndpt, err := store.GetFHIREndpointUsingURLAndListSource(ctx, endpt1.FHIRPatientFacingURI, endpt1.ListSource)
+	th.Assert(t, err == nil, err)
+	th.Assert(t, storedEndpt != nil, "Did not store first product as expected")
+	storedEndpt, err = store.GetFHIREndpointUsingURLAndListSource(ctx, endpt2.FHIRPatientFacingURI, endpt2.ListSource)
+	th.Assert(t, err == nil, err)
+	th.Assert(t, storedEndpt != nil, "Did not store first product as expected")
```
Comment (bmath10) [inline endpointmanager/pkg/fhirendpointquerier/populatefhirendpoints_integration_test.go:271]: Outdated comment

**Comment by bmath10** [comment]
I ran through your instructions and for some reason I don't seem to have any deleted endpoints in the `fhir_endpoints_info_history` table? There were definitely 10 fewer endpoints in the `fhir_endpoints` table, but i don't think `fhir_endpoints_info` decreased. Any ideas?

**Comment by mayerm94** [comment]
> I ran through your instructions and for some reason I don't seem to have any deleted endpoints in the `fhir_endpoints_info_history` table? There were definitely 10 fewer endpoints in the `fhir_endpoints` table, but i don't think `fhir_endpoints_info` decreased. Any ideas?

This might be a result of the logic added after my last comment -- if the 10 endpoints that were removed between EndpointSources and EndpointSources_1 are _also_ in another list (most likely the CareEvolution list) then the _info histories will not be deleted.

File: e2e/integration_tests/data_flow_test.go [hunk 3]
```diff
@@ -436,9 +440,76 @@ func Test_RetrieveCapabilityStatements(t *testing.T) {
...
+	ctx := context.Background()
+	endpointlinker.LinkAllOrgsAndEndpoints(ctx, store, false)
+
+	var link_count int
+	endpoint_orgs_row := store.DB.QueryRow("SELECT COUNT(*) FROM endpoint_organization;")
+	err = endpoint_orgs_row.Scan(&link_count)
+	failOnError(err)
+	if link_count != expected_link_count {
+		t.Fatalf("endpoint_organization should still have %d links after update", link_count)
+	}
```
[the remark this answers is not in the document]
Comment (karenzheung) [reply e2e/integration_tests/data_flow_test.go:469]: expected_link_count is 38 which is the same as in Test_EndpointLinksAreAvailable. If an endpoint is removed from a listsource, when that listsource is updated those endpoints should be removed from fhir_endpoints but endpoint_organization to still show the mapping for previous endpoints.

**Comment by karenzheung** [comment]
> I ran through your instructions and for some reason I don't seem to have any deleted endpoints in the `fhir_endpoints_info_history` table? There were definitely 10 fewer endpoints in the `fhir_endpoints` table, but i don't think `fhir_endpoints_info` decreased. Any ideas?

I was testing using just the CareEvolution listsource. If you comment out lines 8-9 (cerner and epic listsource) in populatedb.sh for both the first and second run then you should see deleted endpoints in fhir_endpoints_info

File: endpointmanager/pkg/fhirendpointquerier/populatefhirendpoints_integration_test.go (hunk 1 above)
Comment (karenzheung) [reply endpointmanager/pkg/fhirendpointquerier/populatefhirendpoints_integration_test.go:258]: Merge test and added another one

**Pull request merged**